lib/elastic_apm/transport/base.rb

```diff
@@ -8,90 +8,156 @@
-        warn 'Queue is full (%i items), skipping…', config.api_buffer_size
+        warn '%s: Queue is full (%i items), skipping…', pid_str, config.api_buffer_size
+        nil
+      rescue Exception => e
+        error '%s: Failed adding to the transport queue: %p', pid_str, e.inspect
         nil
       end
+      # rubocop:enable Metrics/MethodLength, Metrics/LineLength
 
       def add_filter(key, callback)
         @filters.add(key, callback)
       end
 
       private
 
+      def pid_str
+        format('[PID:%s]', Process.pid)
+      end
+
+      def ensure_watcher_running
+        # pid has changed == we've forked
+        return if @pid == Process.pid
+
+        @watcher_mutex.synchronize do
+          return if @pid == Process.pid
+          @pid = Process.pid
+
+          @watcher = Concurrent::TimerTask.execute(
+            execution_interval: WATCHER_EXECUTION_INTERVAL,
+            timeout_interval: WATCHER_TIMEOUT_INTERVAL
+          ) { ensure_worker_count }
+        end
+      end
+
       def ensure_worker_count
-        missing = config.pool_size - @workers.length
-        return unless missing > 0
+        return if all_workers_alive?
+
+        @worker_mutex.synchronize do
+          return if stopped.true?
+
+          @workers.map! do |thread|
+            next thread if thread&.alive?
 
-        info 'Booting %i workers', missing
-        missing.times { boot_worker }
+            boot_worker
+          end
+        end
+      end
+
+      def all_workers_alive?
+        !!workers.all? { |t| t&.alive? }
       end
 
-      # rubocop:disable Metrics/MethodLength
       def boot_worker
-        worker = Worker.new(
-          config,
-          queue,
-          serializers: @serializers,
-          filters: @filters
-        )
-
-        @workers.push worker
-
-        @pool.post do
-          worker.work_forever
-          @workers.delete(worker)
+        debug '%s: Booting worker...', pid_str
+
+        Thread.new do
+          Worker.new(
+            config, queue,
+            serializers: @serializers,
+            filters: @filters
+          ).work_forever
         end
       end
-      # rubocop:enable Metrics/MethodLength
 
+      # rubocop:disable Metrics/MethodLength
       def stop_workers
-        return unless @pool.running?
+        debug '%s: Stopping workers', pid_str
 
-        debug 'Stopping workers'
         send_stop_messages
 
-        debug 'Shutting down pool'
-        @pool.shutdown
-
-        return if @pool.wait_for_termination(5)
+        @worker_mutex.synchronize do
+          workers.each do |thread|
+            next if thread.nil?
+            next if thread.join(WORKER_JOIN_TIMEOUT)
+
+            debug(
+              '%s: Worker did not stop in %ds, killing...',
+              pid_str, WORKER_JOIN_TIMEOUT
+            )
+            thread.kill
+          end
+        end
 
-        warn "Worker pool didn't close in 5 secs, killing ..."
-        @pool.kill
+        @workers.clear
```

Comment:
I don't see a scenario where this would be necessary with the current code. 
Following scenarios are possible:
* T2 allocates mutex in `ensure_worker_count` _while_ T1 sends stop message, T1 allocates mutex in `stop_workers`, T2 runs `ensure_worker_count` again and allocates mutex, but now returns as `stopped=true` _while_ T1 runs `workers.clear`. 
* T1 sends stop message, T2 allocates mutex in `ensure_worker_count`but returns as `stopped=true`, T1 stops and clears the workers
* T1 sends stop message, T1 allocates mutex in `stop_workers`, T2 runs `ensure_worker_count`  and allocates mutex, but returns as `stopped=true` _while_ T1 runs `workers.clear`.

Reply:
This broke my brain a little bit :)

What I meant was this scenario:

- the watcher/timer thread is running, enters `ensure_worker_count`
- at the same time, the transport's `stop` method is called

  - this initiates a shutdown of the watcher, but (this is where I could be wrong) it won't wait for the timer to terminate
  - `stop` proceeds to call `stop_workers`, which calls `@workers.clear` without holding a mutex
  - timer thread in `ensure_worker_count` concurrently calls `all_workers_alive?`